File: src/info/meodinger/LabelPlusFX/Component/CTreeMenu.java
package info.meodinger.LabelPlusFX.Component;

import info.meodinger.LabelPlusFX.State;
import info.meodinger.LabelPlusFX.I18N;
import info.meodinger.LabelPlusFX.Type.TransFile;
import info.meodinger.LabelPlusFX.Type.TransLabel;
import info.meodinger.LabelPlusFX.Util.CColor;
import info.meodinger.LabelPlusFX.Util.CDialog;
import info.meodinger.LabelPlusFX.Util.CString;

import javafx.collections.ObservableList;
import javafx.geometry.Pos;
import javafx.scene.control.*;
import javafx.scene.layout.HBox;
import javafx.scene.paint.Color;
import javafx.scene.shape.Circle;

import java.util.List;
import java.util.Optional;

/**
 * @author Meodinger
 * Date: 2021/5/29
 * Location: info.meodinger.LabelPlusFX.Component
 */
public class CTreeMenu {

    public final TargetMenu rootMenu;
    public final TargetMenu groupMenu;
    public final TargetMenu labelMenu;
    public final TreeMenu treeMenu;

    public CTreeMenu(State state) {
        rootMenu = new RootMenu(state);
        groupMenu = new GroupMenu(state);
        labelMenu = new LabelMenu(state);
        treeMenu = new TreeMenu(state);
    }

    public abstract static class TargetMenu extends ContextMenu {

        private final State state;
        private TreeItem<String> item;

        public TargetMenu(State state) {
            super();
            this.state = state;
            render();
        }

        public void setItem(TreeItem<String> item) {
            this.item = item;
        }

        public TreeItem<String> getItem() {
            return item;
        }

        public State getState() {
            return state;
        }

        protected abstract void render();
        public abstract void init(TreeItem<String> item);
    }
    public static class RootMenu extends TargetMenu {

        private MenuItem addGroup;

        private Dialog<TransFile.MeoTransFile.Group> dialog;
        private CColorPicker colorPicker;
        private TextField nameField;

        public RootMenu(State state) {
            super(state);
        }

        @Override
        protected void render() {
            addGroup = new MenuItem(I18N.MENU_ITEM_ADD_GROUP);
            getItems().add(addGroup);

            colorPicker = new CColorPicker();
            nameField = new TextField();
            nameField.setTextFormatter(new TextFormatter<String>(change -> {
                change.setText(change.getText().trim().replaceAll(" ", "_"));
                return change;
            }));

            dialog = new Dialog<>();
            dialog.getDialogPane().getButtonTypes().addAll(
                    new ButtonType(I18N.SUBMIT, ButtonBar.ButtonData.OK_DONE),
                    new ButtonType(I18N.CANCEL, ButtonBar.ButtonData.CANCEL_CLOSE)
            );
            HBox hBox = new HBox(nameField, colorPicker);
            hBox.setAlignment(Pos.CENTER);
            dialog.getDialogPane().setContent(hBox);
            dialog.setResultConverter((type) -> {
                if (type.getButtonData() == ButtonBar.ButtonData.OK_DONE){
                    return new TransFile.MeoTransFile.Group(
                            nameField.getText().trim().replaceAll(" ", "_"),
                            CColor.toHex(colorPicker.getValue())
                    );
                }
                return null;
            });
            colorPicker.hide();
        }

        public void init(TreeItem<String> item) {
            setItem(item);

            if (item == null) {
                addGroup.setOnAction(null);
                return;
            }

            addGroup.setOnAction(e -> {
                TreeItem<String> rootItem = getItem();
                State state = getState();

                int newGroupId = state.getGroupCount();
                nameField.setText(String.format(I18N.FORMAT_NEW_GROUP_NAME, newGroupId + 1));
                Color color;
                if (newGroupId < 9) {
                    color = Color.web(TransFile.MeoTransFile.DEFAULT_COLOR_LIST[newGroupId]);
                } else {
                    color = Color.WHITE;
                }
                colorPicker.setValue(color);

                dialog.setTitle(I18N.DIALOG_TITLE_ADD_GROUP);
                dialog.setHeaderText(I18N.DIALOG_CONTENT_ADD_GROUP);
                dialog.setResult(null);

                Optional<TransFile.MeoTransFile.Group> result = dialog.showAndWait();
                if (result.isPresent()) {
                    TransFile.MeoTransFile.Group group = result.get();

                    // Edit data
                    state.getGroups().add(group);
                    // Update view
                    state.getControllerAccessor().updateGroupList();
                    state.getControllerAccessor().addLabelLayer();
                    rootItem.getChildren().add(new TreeItem<>(group.name, new Circle(8, Color.web(group.color))));
                    // Mark change
                    state.setChanged(true);
                }
            });
        }
    }
    public static class GroupMenu extends TargetMenu {

        private MenuItem rename, changeColor, delete;
        private CColorPicker picker;

        public GroupMenu(State state) {
            super(state);
        }

        @Override
        protected void render() {
            rename = new MenuItem(I18N.MENU_ITEM_RENAME);
            changeColor = new MenuItem();
            delete = new MenuItem(I18N.MENU_ITEM_DELETE);

            getItems().add(rename);
            getItems().add(changeColor);
            getItems().add(new SeparatorMenuItem());
            getItems().add(delete);

            picker = new CColorPicker();
            picker.valueProperty().addListener((observable, oldValue, newValue) -> changeColor.setText(CColor.toHex(newValue)));
            picker.setPrefSize(40, 20);
            changeColor.setGraphic(picker);
        }

        @Override
        public void init(TreeItem<String> item) {
            setItem(item);

            if (item == null) {
                rename.setOnAction(null);
                picker.setOnAction(null);
                delete.setOnAction(null);
                return;
            }

            rename.setOnAction(e -> {
                TreeItem<String> groupItem = getItem();
                State state = getState();

                Optional<String> result = CDialog.showInput(state.stage, I18N.DIALOG_TITLE_RENAME, I18N.DIALOG_CONTENT_RENAME, groupItem.getValue());

                if (result.isPresent() && !CString.isBlank(result.get())) {
                    String name = result.get().trim().replaceAll(" ", "_");
                    int groupId = state.getGroupIdByName(groupItem.getValue());

                    if (state.isLPFile() && state.getGroupNames().contains(name)) {
                        CDialog.showAlert(I18N.EXPORTER_SAME_GROUP_NAME);
                        return;
                    }

                    // Edit data
                    state.getGroupAt(groupId).name = name;
                    // Update view
                    state.getControllerAccessor().updateGroupList();
                    groupItem.setValue(name);
                    // Mark change
                    state.setChanged(true);
                }
            });

            picker.setValue((Color) ((Circle) getItem().getGraphic()).getFill());
            picker.setOnAction(e -> {
                State state = getState();
                Color color = ((ColorPicker) e.getSource()).getValue();
                int groupId = state.getGroupIdByName(getItem().getValue());

                // Edit data
                state.getGroupAt(groupId).color = CColor.toHex(color);
                // Update view
                ((Circle) getItem().getGraphic()).setFill(color);
                state.getControllerAccessor().updateLabelLayer(groupId);
                // Mark change
                state.setChanged(true);

            });
            changeColor.setText(CColor.toHex(picker.getValue()));


            int labelCount = 0, thisGroupId = getState().getGroupIdByName(getItem().getValue());
            for (List<TransLabel> labels : getState().getTransMap().values()) {
                for (TransLabel label : labels) {
                    if (label.getGroupId() == thisGroupId) labelCount++;
                }
            }
            delete.setDisable(labelCount != 0);
            delete.setOnAction(e -> {
                TreeItem<String> groupItem = getItem();
                State state = getState();
                int groupId = state.getGroupIdByName(groupItem.getValue());

                // Edit data
                for (List<TransLabel> labels : state.getTransMap().values()) for (TransLabel label : labels)
                    if (label.getGroupId() >= groupId) label.setGroupId(label.getGroupId() - 1);
                state.getGroups().remove(groupId);
                // Update view
                groupItem.getParent().getChildren().remove(groupItem);
                state.getControllerAccessor().updateGroupList();
                state.getControllerAccessor().removeLabelLayer(groupId);
                // Mark change
                state.setChanged(true);
            });
        }
    }
    public static class LabelMenu extends TargetMenu {

        private MenuItem editGroup, delete;

        public LabelMenu(State state) {
            super(state);
        }

        @Override
        protected void render() {
            editGroup = new MenuItem(I18N.MENU_ITEM_MOVE_TO);
            delete = new MenuItem(I18N.MENU_ITEM_DELETE);

            getItems().add(editGroup);
            getItems().add(new SeparatorMenuItem());
            getItems().add(delete);
        }

        @Override
        public void init(TreeItem<String> item) {
            setItem(item);

            if (item == null) {
                editGroup.setOnAction(null);
                delete.setOnAction(null);
                return;
            }

            editGroup.setOnAction(e -> {
                CTreeItem labelItem = (CTreeItem) getItem();
                State state = getState();
                int prevGroupId = labelItem.meta.getGroupId();
                Optional<String> result = CDialog.showChoice(state.stage, I18N.DIALOG_TITLE_MOVE_TO, I18N.DIALOG_CONTENT_MOVE_TO, state.getGroupNames());

                if (result.isPresent()) {
                    // Edit data
                    labelItem.setGroupName(result.get());
                    // Update view
                    state.getControllerAccessor().updateLabelLayer(prevGroupId);
                    state.getControllerAccessor().updateLabelLayer(labelItem.meta.getGroupId());
                    state.getControllerAccessor().updateTree();
                    // Mark change
                    state.setChanged(true);
                }
            });

            /*
             * Please also edit `CImagePane#handleLabelMode#SECONDARY`
             */
            delete.setOnAction(e -> {
                CTreeItem labelItem = (CTreeItem) getItem();
                State state = getState();

                Optional<ButtonType> result = CDialog.showConfirm(I18N.DIALOG_TITLE_DELETE_LABEL, I18N.DIALOG_CONTENT_DELETE_LABEL, labelItem.getValue());

                if (result.isPresent() && result.get() == ButtonType.YES) {

                    // Edit data
                    for (TransLabel l : state.getLabelsNow())
                        if (l.getIndex() > labelItem.meta.getIndex()) l.setIndex(l.getIndex() - 1);
                    state.getLabelsNow().remove(labelItem.meta);
                    // Update view
                    state.getControllerAccessor().updateLabelLayer(labelItem.meta.getGroupId()); // Also update positions
                    state.getControllerAccessor().updateTree();
                    // Mark change
                    state.setChanged(true);
                }
            });
        }
    }

    public static class TreeMenu extends ContextMenu {

        private final State state;
        private MenuItem editGroup, delete;
        private TreeView<String> view;

        public TreeMenu(State state) {
            super();
            this.state = state;
            render();
        }

        private void render() {
            editGroup = new MenuItem(I18N.MENU_ITEM_MOVE_TO);
            delete = new MenuItem(I18N.MENU_ITEM_DELETE);

            getItems().add(editGroup);
            getItems().add(new SeparatorMenuItem());
            getItems().add(delete);
        }

        public void init(TreeView<String> treeView) {
            view = treeView;

            editGroup.setOnAction(e -> {
                ObservableList<TreeItem<String>> selectedItems = view.getSelectionModel().getSelectedItems();

                Optional<String> result = CDialog.showChoice(state.stage, I18N.DIALOG_TITLE_MOVE_TO, I18N.DIALOG_CONTENT_MOVE_TO, state.getGroupNames());

                // Edit data
                if (result.isPresent()) {
                    String name = result.get();
                    for (TreeItem<String> item : selectedItems) {
                        ((CTreeItem) item).setGroupName(name);
                    }
                }
                // Update view
                state.getControllerAccessor().updatePane();
                state.getControllerAccessor().updateTree();
                // Mark change
                state.setChanged(true);
            });

            delete.setOnAction(e -> {
                ObservableList<TreeItem<String >> selectedItems = view.getSelectionModel().getSelectedItems();

                Optional<ButtonType> result = CDialog.showConfirm(I18N.DIALOG_TITLE_DELETE_LABEL, I18N.DIALOG_CONTENT_DELETE_LABELS);

                if (result.isPresent() && result.get() == ButtonType.YES) {
                    // Edit data
                    for (TreeItem<String> item : selectedItems) {
                        TransLabel label = ((CTreeItem) item).meta;
                        for (TransLabel l : state.getLabelsNow())
                            if (l.getIndex() > label.getIndex()) l.setIndex(l.getIndex() - 1);
                        state.getLabelsNow().remove(label);
                    }
                    // Update view
                    state.getControllerAccessor().updatePane(); // Also update positions
                    state.getControllerAccessor().updateTree();
                    // Mark change
                    state.setChanged(true);
                }
            });
        }

        public void update() {
            editGroup.setDisable(true);
            delete.setDisable(true);

            ObservableList<TreeItem<String>> selectedItems = view.getSelectionModel().getSelectedItems();

            if (selectedItems.size() == 0) return;
            for (TreeItem<String> item : selectedItems) {
                if (item.getClass() != CTreeItem.class) {
                    return;
                }
            }

            editGroup.setDisable(false);
            delete.setDisable(false);
        }
    }


}

File: src/main/kotlin/info/meodinger/lpfx/component/CTreeMenu.kt
package info.meodinger.lpfx.component

import info.meodinger.lpfx.State
import info.meodinger.lpfx.options.Settings
import info.meodinger.lpfx.type.TransGroup
import info.meodinger.lpfx.util.color.toHex
import info.meodinger.lpfx.util.dialog.*
import info.meodinger.lpfx.util.resource.I18N
import info.meodinger.lpfx.util.resource.get

import javafx.geometry.Pos
import javafx.scene.control.*
import javafx.scene.layout.HBox
import javafx.scene.paint.Color
import javafx.scene.shape.Circle

/**
 * Author: Meodinger
 * Date: 2021/7/29
 * Location: info.meodinger.lpfx.component
 */
object CTreeMenu {

    private val GroupNameTextFormatter = TextFormatter<String> { change ->
        change.text = change.text
            .trim()
            .replace(" ", "_")
            .replace("|", "_")
        change
    }

    abstract class TargetMenu : ContextMenu() {
        abstract fun redirectTo(target: TreeItem<String>?)
    }
    val rootMenu = object : TargetMenu() {

        private val addGroupField = TextField()
        private val addGroupPicker = CColorPicker()
        private val addGroupItem = MenuItem(I18N["context.add_group"])
        private val addGroupDialog = Dialog<TransGroup>()
        private val addGroupAction = { rootItem: TreeItem<String> ->
            val newGroupId = State.transFile.groupList.size
            val newColor = if (newGroupId >= 9) Color.WHITE
            else Color.web(Settings[Settings.DefaultColorList].asList()[newGroupId])

            addGroupField.text = String.format(I18N["context.add_group.dialog.format"], newGroupId + 1)
            addGroupPicker.value = newColor
            addGroupDialog.result = null
            addGroupDialog.showAndWait().ifPresent { newGroup ->
                // Edit data
                State.transFile.groupList.add(newGroup)
                // Update view
                State.accessor.updateGroupList()
                State.accessor.addLabelLayer()
                rootItem.children.add(TreeItem(newGroup.name, Circle(8.0, Color.web(newGroup.color))))
                // Mark change
                State.isChanged = true
            }
        }

        init {
            addGroupField.textFormatter = GroupNameTextFormatter
            addGroupPicker.hide()
            addGroupDialog.title = I18N["context.add_group.dialog.title"]
            addGroupDialog.headerText = I18N["context.add_group.dialog.header"]
            addGroupDialog.dialogPane.content = HBox(addGroupField, addGroupPicker).also { it.alignment = Pos.CENTER }
            addGroupDialog.dialogPane.buttonTypes.addAll(ButtonType.APPLY, ButtonType.CANCEL)
            addGroupDialog.setResultConverter {
                if (it == ButtonType.APPLY)
                    TransGroup(addGroupField.text, addGroupPicker.value.toHex())
                else
                    null
            }

            items.add(addGroupItem)
        }

        override fun redirectTo(target: TreeItem<String>?) {
            if (target == null) {
                addGroupItem.onAction = null
            } else {
                addGroupItem.setOnAction { addGroupAction(target) }
            }
        }

    }
    val groupMenu = object : TargetMenu() {

        private val renameItem = MenuItem(I18N["context.rename_group"])
        private val renameAction = { groupItem: TreeItem<String> ->
            showInput(
                State.stage,
                I18N["context.rename_group.dialog.title"],
                I18N["context.rename_group.dialog.header"],
                groupItem.value,
                TextFormatter<String> { change ->
                    change.text = change.text
                        .trim()
                        .replace(" ", "_")
                        .replace("|", "_")
                    change
                }
            ).ifPresent { newName ->
                if (newName.isBlank()) return@ifPresent
                for (group in State.transFile.groupList) {
                    if (group.name == newName) showError(I18N["context.rename_group.error.same_name"])
                    return@ifPresent
                }

                val groupId = State.getGroupIdByName(groupItem.value)

                // Edit data
                State.transFile.getTransGroupAt(groupId).name = newName
                // Update view
                State.accessor.updateGroupList()
                groupItem.value = newName
                // Mark change
                State.isChanged = true
            }
        }

        private val changeColorItem = MenuItem()
        private val changeColorPicker = CColorPicker()
        private val changeColorAction = { groupItem: TreeItem<String> ->
            val newColor = changeColorPicker.value
            val groupId = State.getGroupIdByName(groupItem.value)

            // Edit data
            State.transFile.getTransGroupAt(groupId).color = newColor.toHex()
            // Update view
            (groupItem.graphic as Circle).fill = newColor
            // Mark change
            State.isChanged = true
        }

        private val deleteItem = MenuItem(I18N["context.delete_group"])
        private val deleteAction = { groupItem: TreeItem<String> ->
            val groupId = State.getGroupIdByName(groupItem.value)

            // Edit data
            for (labels in State.transFile.transMap.values) for (label in labels)
                if (label.groupId >= groupId) label.groupId = label.groupId - 1
            State.transFile.groupList.removeIf { it.name == groupItem.value }
            // Update view
            groupItem.parent.children.remove(groupItem)
            State.accessor.updateGroupList()
            State.accessor.removeLabelLayer(groupId)
            // Mark change
            State.isChanged = true
        }

        init {
            changeColorPicker.valueProperty().addListener { _, _, newValue ->
                changeColorItem.text = newValue.toHex()
            }
            changeColorPicker.setPrefSize(40.0, 20.0)
            changeColorItem.graphic = changeColorPicker

            items.add(renameItem)
            items.add(changeColorItem)
            items.add(SeparatorMenuItem())
            items.add(deleteItem)
        }

        override fun redirectTo(target: TreeItem<String>?) {
            if (target == null) {
                renameItem.onAction = null
                changeColorItem.onAction = null
                deleteItem.onAction = null
            } else {
                renameItem.setOnAction { renameAction(target) }

                changeColorPicker.value = (target.graphic as Circle).fill as Color
                changeColorItem.text = changeColorPicker.value.toHex()
                changeColorPicker.setOnAction { changeColorAction(target) }

                deleteItem.isDisable = false
                val thisGroupId = State.getGroupIdByName(target.value)
                outer@ for (labels in State.transFile.transMap.values) {
                    for (label in labels) {
                        if (label.groupId == thisGroupId) {
                            deleteItem.isDisable = true
                            break@outer
                        }
                    }
                }
                deleteItem.setOnAction { deleteAction(target) }
            }
        }

    }
    val labelMenu = object : TargetMenu() {

        private val moveToItem = MenuItem(I18N["context.move_to"])
        private val moveToAction = { item: TreeItem<String> ->
            val labelItem = item as CTreeItem
            val groupNameList = ArrayList<String>()
            for (group in State.transFile.groupList) groupNameList.add(group.name)

            showChoice(
                State.stage,
                I18N["context.move_to.title"],
                I18N["context.move_to.header"],
                groupNameList
            ).ifPresent { newGroupName ->
                // Edit data
                labelItem.groupId = State.getGroupIdByName(newGroupName)
                // Update view
                State.accessor.updateTree()
                // Mark change
                State.isChanged = true
            }
        }

        private val deleteItem = MenuItem(I18N["context.delete_label"])
        private val deleteAction = { item: TreeItem<String> ->
            val labelItem = item as CTreeItem
            val result = showConfirm(
                I18N["context.delete_label.dialog.title"],
                I18N["context.delete_label.dialog.header"],
                labelItem.value
            )

            if (result.isPresent && result.get() == ButtonType.YES) {
                // Edit data
                for (label in State.transFile.getTransLabelListOf(State.currentPicName)) {
                    if (label.index > labelItem.index) {
                        label.index = label.index - 1
                    }
                }
                State.transFile.getTransLabelListOf(State.currentPicName).remove(labelItem.meta)
                // Update view
                State.accessor.updateTree()
                // Mark change
                State.isChanged = true
            }
        }

        init {
            items.add(moveToItem)
            items.add(SeparatorMenuItem())
            items.add(deleteItem)
        }

        override fun redirectTo(target: TreeItem<String>?) {
            if (target == null) {
                moveToItem.onAction = null
                deleteItem.onAction = null
            } else {
                moveToItem.setOnAction { moveToAction(target) }
                deleteItem.setOnAction { deleteAction(target) }
            }
        }

    }

    abstract class TreeMenu : ContextMenu() {
        abstract fun init(view: TreeView<String>?)
        abstract fun update()
    }
    val treeMenu = object : TreeMenu() {

        private var view: TreeView<String>? = null

        private val moveToItem = MenuItem(I18N["context.move_to"])
        private val moveToAction = { view: TreeView<String> ->
            val selectedItems = view.selectionModel.selectedItems
            val groupNameList = ArrayList<String>()
            for (group in State.transFile.groupList) groupNameList.add(group.name)

            showChoice(
                State.stage,
                I18N["context.move_to.title"],
                I18N["context.move_to.header"],
                groupNameList
            ).ifPresent { newGroupName ->
                val newGroupId = State.getGroupIdByName(newGroupName)

                // Edit data
                for (item in selectedItems) {
                    (item as CTreeItem).groupId = newGroupId
                }
                // Update view
                State.accessor.updateTree()
                // Mark change
                State.isChanged = true
            }
        }

        private val deleteItem = MenuItem(I18N["context.delete_label"])
        private val deleteAction = { view: TreeView<String> ->
            val selectedItems = view.selectionModel.selectedItems
            val result = showConfirm(
                I18N["context.delete_label.dialog.title"],
                null,
                I18N["context.delete_label.dialog.content.pl"],
            )

            if (result.isPresent && result.get() == ButtonType.YES) {
                // Edit data
                for (item in selectedItems) {
                    val label = (item as CTreeItem).meta
                    for (l in State.transFile.getTransLabelListOf(State.currentPicName)) {
                        if (l.index > label.index)
                            l.index = l.index - 1
                    }
                    State.transFile.getTransLabelListOf(State.currentPicName).remove(label)
                }
                // Update view
                State.accessor.updateTree()
                // Mark change
                State.isChanged = true
            }
        }

        init {
            items.add(moveToItem)
            items.add(SeparatorMenuItem())
            items.add(deleteItem)
        }

        override fun init(view: TreeView<String>?) {
            this.view = view

            if (view == null) {
                moveToItem.onAction = null
                deleteItem.onAction = null
            } else {
                moveToItem.setOnAction { moveToAction(view) }
                deleteItem.setOnAction { deleteAction(view) }
            }
        }

        override fun update() {
            moveToItem.isDisable = true
            deleteItem.isDisable = true

            if (view != null) {
                val selectedItems = view!!.selectionModel.selectedItems

                if (selectedItems.size == 0) return
                for (item in selectedItems) {
                    if (item.javaClass != CTreeItem::class.java) {
                        return
                    }
                }
            }

            moveToItem.isDisable = false
            deleteItem.isDisable = false
        }

    }

}